git-remote-ipfs/src/Network/IPFS/Git/RemoteHelper.hs

```diff
@@ -175,33 +188,49 @@ processPush _ localRef remoteRef = do
                 liftEitherRH . first CidError $
                     cidToRef @Git.SHA1 localRefCid
             logDebug $ "sha " <> Text.pack (show sha)
-            dir <- Git.gitRepoPath <$> Git.getGit
-            git . liftIO $ Git.looseRead dir sha
+            git $ do
+                g <- Git.getGit
+                liftIO $ Git.getObject_ g sha True
 
         let raw = Git.looseMarshall obj
 
-        logDebug $ "BlockPut: " <> decodeUtf8With lenientDecode (L.toStrict raw)
-        blockCid <- do
-            cid <- ipfs $ putBlock raw
-            -- check 'res' CID matches SHA
-            when (localRefCid /= cid) $
-                throwRH
-                    . CidError
-                    $ sfmt ("CID mismatch: expected `" % fcid % "`, actual `" % fcid % "`")
-                           localRefCid
-                           cid
-            pure cid
-
-        logDebug $ fmt ("blockCid: " % fcid) blockCid
-
-        -- if loose object > 2048k, create object + link block to it
-        when (L.length raw > 2048000) $
-            void $ linkedObject root ("objects/" <> cidToText blockCid) raw
+        blkCid <- ipfs $ putBlock raw
+        when (localRefCid /= blkCid) $
+            throwRH $ HashError (CidMismatch localRefCid blkCid)
+
+        -- If the object exceeds the maximum block size, bitswap won't replicate
+        -- the block. To work around this, we create a regular object and link
+        -- it to the root object as @objects/<block CID>@.
+        --
+        -- As suggested by
+        -- <https://github.com/ipfs-shipyard/git-remote-ipld/issues/12>, objects
+        -- can potentially be deduplicated by storing the data separate from the
+        -- header. This only makes sense for git blobs, so we don't bother for
+        -- other object types.
+        when (L.length raw > fromIntegral clientMaxBlockSize) $
+            let
+                linkName = "objects/" <> cidToText blkCid
```

Comment:
Note that we still have to create the block, even though we could just use `localRefCid`. I have no idea why, but fetch otherwise does not make any progress.